src/org/etools/j1939_84/controllers/part12/Part12Step11Controller.java

```diff
@@ -73,9 +82,162 @@ protected void run() throws Throwable {
+        });
+        getDataRepository().getObdModuleAddresses().forEach(moduleAddress -> {
+            getDm58AndVerifyData(moduleAddress);
+        });
+    }
+
+    private void getDm58AndVerifyData(int moduleAddress) {
+
+        // 6.12.11.4 Actions3:
+        // a. DS DM7 with TID 245 (for DM58) using FMI 31 for first SP identified as not supporting DM58 in a DM24
+        // response In step 6.1.4.1 to the SP’s respective OBD ECU. (Use of an SP that supports test results is
+        // preferred when available).
+        var nonRatFaultSps = getDataRepository().getObdModules()
+                                                .stream()
+                                                .filter(module -> module.getSourceAddress() == moduleAddress)
+                                                .flatMap(m -> m.getSupportedSPNs().stream())
+                                                .filter(supported -> !supported.supportsRationalityFaultData())
+                                                .collect(Collectors.toList());
+        if (nonRatFaultSps.isEmpty()) {
+            getListener().onResult("6.12.11.4.a - No SPs found that do NOT indicate support for DM58 in the DM24 response from "
+                    + Lookup.getAddressName(moduleAddress));
+        } else {
+            int requestSpn = nonRatFaultSps.stream()
+                                           .filter(SupportedSPN::supportsScaledTestResults)
+                                           .findFirst()
+                                           .orElseGet(() -> nonRatFaultSps.get(0))
+                                           .getSpn();
+
+            var packet = getCommunicationsModule().requestDM58(getListener(), moduleAddress, requestSpn)
+                                                  .requestResult()
+                                                  .getAcks()
+                                                  .stream()
+                                                  .findFirst()
+                                                  .orElse(null);
+
+            // 6.12.11.5 Fail/Warn criteria3:
+            // a. Fail if a NACK is not received
+            if (packet == null || packet.getResponse() != AcknowledgmentPacket.Response.NACK) {
+                addFailure("6.12.11.5.a - NACK not received for DM7 PG from OBD ECU "
+                        + Lookup.getAddressName(moduleAddress) + " for spn " + requestSpn);
+            }
+        }
+    }
+
+    private boolean areUnusedBytesPaddedWithFFh(DM58RationalityFaultSpData packet) {
+        Slot slot = J1939DaRepository.findSlot(packet.getSpn().getSlot().getId(), packet.getSpn().getId());
+
+        int slotLength = slot.getByteLength();
+        int spnLength = packet.getSpnDataBytes().length;
+
+        byte[] paddingBytes;
+
+        switch (slotLength) {
+            case 1:
+            case 3:
+                paddingBytes = Arrays.copyOfRange(packet.getSpnDataBytes(), slotLength, spnLength);
```

Comment:
This can’t be correct. For 3 bytes, you should be using bytes 0, 1, 2.

Reply:
This still needs addressed

Reply:
You are incorrect, it should only be packet.getSpnDataBytes()[3].  The variable is labeled "paddingBytes" because I am only collecting the bytes used to PAD.